dashboard/lib/sample_data.rb

```diff
@@ -0,0 +1,181 @@
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: "csp1-2018", num_students: 30,
+      use_imperfect_results: false}
+      )
+
+    create_section ({teacher: teacher, name: "CSP 2 (Large)",
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: "csp1-2018", num_students: 150,
+      use_imperfect_results: false}
+      )
+  end
+
+  # will delete the teacher and all of the teacher's sections and students
+  # and recreate.
+  def self.create_teacher(email, name)
+    # Delete any existing test data
+    user = User.find_by_email_or_hashed_email(email)
+    unless user.nil?
+      user.sections.each do |section|
+        # Delete all students in each section
+        section.followers.each do |follower|
+          student_user = User.find_by_id(follower.student_user_id)
+          raise "Not a test user" unless student_user.name.include? "Codeberg"
+          UserGeo.where(user_id: student_user.id).destroy_all
+          student_user.really_destroy!
+        end
+        # Delete each section
+        section.really_destroy!
+      end
+      UserGeo.where(user_id: user.id).destroy_all
+      # Delete the existing test teacher
+      raise "Not a test user" unless user.name.include? "Codeberg"
+      user.really_destroy!
+    end
+    # Create the test teacher
+    create :teacher, {email: email, name: name,
+      password: TEST_ACCOUNT_PASSWORD, terms_of_service_version: 1}
+  end
+
+  # Generate a random gender choice with reasonable distributions
+  def self.random_gender
+    val = prng.rand(100)
+    gender = nil
+    case val
+    when 0..29
+      gender = "m"
+    when 30..59
+      gender = "f"
+    when 60..89
+      gender = nil
+    when 90..94
+      gender = "n"
+    when 95..99
+      gender = "o"
+    end
+    gender
+  end
+
+  # Create a test section
+  # Options is a hash with expected values:
+  #  :teacher - User object of teacher
+  #  :name - section name
+  #  :login_type - Section::LOGIN_TYPE_xxx value of login type for section
+  #  :grade - grade for section
+  #  :age_min - minimum age to generate for students
+  #  :age_max_inclusive - max age (inclusive) to generate for students
+  #  :script_name - name of script to assign for section
+  #  :num_students - number of students to generate for section
+  #  :use_imperfect_results - if true, generate some less than perfect
+  #    results. (CSF allows imperfect results, CSD and CSP do not.)
+  def self.create_section(options)
+    script = Script.get_from_cache(options[:script_name])
+    level_count = script.script_levels.count
+
+    # Create the section
+    section = create :section, {user: options[:teacher], name: options[:name],
+      login_type: options[:login_type], grade: options[:grade],
+      script: script}
+
+    # Create students in section
+    (0..options[:num_students] - 1).each do
```

Comment:
Why use an [inclusive range](https://ruby-doc.org/core-2.2.0/Range.html) from 0 to num students - 1, especially when we're not using the iterator? Remove the -1 and either iterate inclusively from 1 to num_students, or exclusively from 0 to num_students.

```
[development] dashboard > num = 5
=> 5
[development] dashboard > (0..num - 1).map {|n| n} # unneessary math
=> [0, 1, 2, 3, 4]
[development] dashboard > (1..num).map {|n| n} # inclusive from 1
=> [1, 2, 3, 4, 5]
[development] dashboard > (0...num).map {|n| n} # exclusive from 0
=> [0, 1, 2, 3, 4]
```

Reply:
Thx, did (1..num)